### ISSUE TITLE
Benchmarking Code

### ISSUE THREAD

**Issue by ben-heil** [greenelab/saged#3]
The code required to compare different models will have a lot of different components, so I'll list the elements here and check them off as they're implemented.

- [x] Create a base Dataset class
- [x] Implement a Dataset class for refinebio data
- [x] Implement extracting data for a given label
- [x] Implement subsetting data based on label
- [x] Implement train/test split and cross-validation, by study instead of by sample

- [x] Create a base Model class (probably using the sklearn API?)
- [x] Implement Models for skl methods
- [ ] Implement Models for some semi-supervised methods
- [ ] Implement non-semi-supervised versions of the same models, or separate unsupervised and supervised steps

- [ ] Implement data simulation/augmentation with VAEs
- [x] Determine a way to parallelize the benchmarking process
- [x] Write the logic to take a dataset and benchmark models on it. Probably build this into the dataset object itself?

**Comment by ben-heil** [comment]
Assorted notes from planning are below.

Benchmark structure:
- One config file that controls run details. Which batch effect correction to use, which datasets to use, which models
- The dataset and models sections contain lists of config file paths
- Each dataset has a config file denoting paths to the files they need and which function to use to initialize them
- Each model has a config file listing its model class (for finding it in models.py) and its hyperparameter values to use
- Each analysis (e.g. performance of models on all classes, performance of models on single class) are their own python script
- The logic for running everything is strung together with snakemake
- Results formats are still up in the air. Probably tsv, but we need to have lists of study ids, models, etc

 ```
# Across all diseases, do we perform better or worse with embeddings                                                                                                                           
Load dataset                                                                                                                                                                                  
Split compendium data into labeled and unlabeled data                                              
Split labeled data via 5FCV                                                                        
                                                                                                    
Get the number of labels and the number of genes.                                                  
Load the supervised model from the config passed in
If an unsupervised model is called for, load the unsupervised model and embed data
Iterate over CV splits:
    train model
    evaluate model
    save results
                                                                                                                                                   
# How are we storing results?                                                                       
# How do we implement early stopping and resuming? Snakemake?                                       
```
